File: rb/three.rb
# The prime factors of 13195 are 5, 7, 13 and 29.

# What is the largest prime factor of the number 600851475143 ?
require "pry"

class Integer
    def factors
        factors = []
        (1..Math.sqrt(self)).each {|num|
            if self % num == 0 && num != 1 && num != self
                factors.push(num) unless factors.include?(num)
                factors.push(self / num) unless factors.include?(self / num)
            end
        }
        factors.sort
    end
    def is_prime?
        self.factors.length == 0 ? true : false
    end
    def largest_prime_factor
        factors = self.factors.reverse
        factors.each { |factor|
            if factor.is_prime?
                return factor
            end
        }
    end
end

puts "#{600851475143.largest_prime_factor} is the largest factor"

File: js/three.js
// The prime factors of 13195 are 5, 7, 13 and 29.

// What is the largest prime factor of the number 600851475143 ?

function largest_prime_factor(num) {
  var factors = get_factors(num);
  var factorsLength = factors.length;
  for(var i = (factorsLength - 1); i >= 0; i--){
    if(is_prime(factors[i])){
      return factors[i];
    }
  }  
}
function get_factors(num){
  var factors = [];
  var max = parseInt(Math.sqrt(num));
  for(var i = 1; i <= max; i++){
    if(num % i == 0){
      if(factors.indexOf(i) < 0){
        factors.push(i);
      }
      if(factors.indexOf(num / i) < 0){
        factors.push(num / i);
      }
    }
  }
  return factors.sort(function(a,b){return a - b;});
}
function is_prime(num){
  if(get_factors(num).length == 2){
    return num;
  }
}
console.log(largest_prime_factor(600851475143));